File: src_jl_demo/LVisP_comparison.jl
#%%imports
using CSV
using DataFrames
using Glob
using NaNStatistics
using Plots
using Random
using Revise

include(joinpath(@__DIR__, "../src/LVisP.jl"))
using .LVisP

#%%definitions
"""
    - function defining a gaussian normaldistribution
"""
function gaussian_pdf(x, mu, sigma)
    f = 1 / (sigma * sqrt(2π)) * exp(-(x - mu)^2 / (2sigma^2))
    return f
end
"""
    - function to define a very simplistic phenomenological LC simulation
"""
function lc_sim(
    t::Vector;
    t_peak::Number, f_peak::Number,
    lambda::Number=1.0,
    stretch0::Number, stretch1::Number, stretch2::Number,
    noiselevel::Number=0.0,
    )::Vector

    f = (gaussian_pdf.(t, t_peak - stretch0/2, stretch1) .+ gaussian_pdf.(t, t_peak + stretch0/2, stretch2))
    f = f_peak .* f ./ maximum(f) .+ noiselevel .* randn(length(t))
    f .*= lambda #wavelength dependent scaling
    return f
end
"""
    - function to evaluate a sin with period `p` and `offset`
"""
function sin_sim(
    t::Real;
    f_peak::Real,
    p::Real, offset::Real=0.,
    noiselevel::Number=0.0
    )::Real
    f = f_peak * sin(t * 2pi/p + offset) .+ noiselevel .* randn()
    return f
end
"""
    - function to generate `nobjects` different dataseries according to `opt` (one of `"lc"`, `"sin"`)
"""
function simulate(
    nobjects::Int=12;
    opt::String="lc"
    )
    x = Array.(eachcol(collect(-20:0.2:100) .* ones(nobjects)'))
    theta_options = 0.4:0.3:4
    ridxs = randperm(length(theta_options))  #for random sampling without replacement
    theta = theta_options[ridxs][1:nobjects]
    
    if opt == "lc"
        t_peak = collect(range(0,40, nobjects)) .* 1
        y = map(i -> lc_sim(x[i]; t_peak=t_peak[i], f_peak=20, lambda=theta[i], stretch0=5, stretch1=15, stretch2=40, noiselevel=1.0), 1:nobjects)
        y_nonoise = map(i -> lc_sim(x[i]; t_peak=t_peak[i], f_peak=20, lambda=theta[i], stretch0=5, stretch1=15, stretch2=40, noiselevel=0.0), 1:nobjects)
    else
        theta .*= 10
        y = map(i -> sin_sim.(x[i]; f_peak=1, p=theta[i], offset=0.0, noiselevel=0.15), 1:nobjects)
        y_nonoise = map(i -> sin_sim.(x[i]; f_peak=1, p=theta[i], offset=0.0, noiselevel=0.0), 1:nobjects)
    end

    df_raw = DataFrame(
        :period=>vcat(repeat(theta, 1, length(x[1]))'...),
        :time=>vcat(x...),
        :amplitude=>vcat(y...),
        :amplitude_e=>NaN,
        :processing=>"raw",
    )
    df_pro = DataFrame(
        :period=>vcat(repeat(theta, 1, length(x[1]))'...),
        :time=>vcat(x...),
        :amplitude=>vcat(y_nonoise...),
        :amplitude_e=>NaN,
        :processing=>"nonoise",

    )
    df = vcat(df_raw, df_pro)

    # p = scatter(
    #     df_raw[!,:time], df_raw[!,:amplitude],
    #     group=df_raw[!,:period],
    #     )
    # plot!(p,
    #     df_pro[!,:time], df_pro[!,:amplitude],
    #     group=df_pro[!,:period],

    # )
    # display(p)

    return df
end


#%%demos

#%%data loading
fnames = Glob.glob("*.csv", joinpath(@__DIR__,"../data/"))
append!(fnames, [joinpath(@__DIR__, "../data/lc_simulated.jl"), joinpath(@__DIR__, "../data/sin_simulated.jl")])    #append pseudo filenames for data generated in this script
println(fnames)
fname = fnames[31]

#deal with on-the-fly data generation (pseudo filenames)
if fname == joinpath(@__DIR__, "../data/lc_simulated.jl")
    df = simulate(10; opt="lc")
elseif fname == joinpath(@__DIR__, "../data/sin_simulated.jl")
    df = simulate(10; opt="sin")
else
    df = DataFrame(CSV.File(fname))
end

parts = split(fname, r"[\_\.]")
survey = parts[end-1]
otype = parts[end-2]

df_raw = subset(df, :processing => p -> p .== "raw")    #raw
df_pro = subset(df, :processing => p -> p .!= "raw")    #processed


theta_raw = sort(unique(df_raw[!,1]))
x_raw = map(x -> Vector(x[!,2]), collect(groupby(df_raw, 1; sort=true)))
y_raw = map(y -> Vector(y[!,3]), collect(groupby(df_raw, 1; sort=true)))
theta_pro = sort(unique(df_pro[!,1]))
x_pro = map(x -> Vector(x[!,2]), collect(groupby(df_pro, 1; sort=true)))
y_pro = map(y -> Vector(y[!,3]), collect(groupby(df_pro, 1; sort=true)))

x_ref = nanminimum.(x_raw)
y_ref = nanminimum.(y_raw)

x_raw = map((x, xr) -> x .- xr, x_raw, x_ref)
x_pro = map((x, xr) -> x .- xr, x_pro, x_ref)
# y_raw = map((y, yr) -> y .- yr, y_raw, y_ref)
# y_pro = map((y, yr) -> y .- yr, y_pro, y_ref)

# y_raw = map(y -> y .- nanminimum(y), y_raw)
# y_pro = map(y -> y .- nanminimum(y), y_pro)

# theta_raw   = theta_raw[1:1]
# theta_pro   = theta_pro[1:1]
# x_raw       = x_raw[    1:1]
# x_pro       = x_pro[    1:1]
# y_raw       = y_raw[    1:1]
# y_pro       = y_pro[    1:1]

# p = scatter(x_raw, y_raw)
# plot!(p, x_pro, y_pro)
# display(p)


#%%get stats
unique_thetas = unique(theta_raw)
nthetas = length(unique_thetas)

display(unique_thetas)
display(nthetas)


#%%get required parameters
thetalims = (minimum(theta_raw), maximum(theta_raw))
xticks = collect(range(floor(minimum(minimum.(x_raw))), ceil(maximum(maximum.(x_raw))), 5))
yticks = collect(range(floor(minimum(minimum.(y_pro))), ceil(maximum(maximum.(y_pro))), 3))

# xticks = (xticks, ["x1 ", "x2 ", "x3 ", "x4 ", "x5 "])
# yticks = (yticks, string.(log10.(yticks)))
# yticks = nothing
# yticks = [17, 21]

# display(thetalims)
# display(xticks)
# display(extrema(x_pro[1]))


#%%plotting
#LVisP
panelsize = pi/(2*nthetas)
LVPC = LVisP.LVisPCanvas(
    thetalims=thetalims, xticks=xticks,
    thetaguidelims=(-pi/2,pi/2), thetaplotlims=(-pi/2+panelsize/2,pi/2-panelsize/2), xlimdeadzone=0.3, panelsize=panelsize,
    thetalabel="Passband Wavelength [nm]", xlabel="\n\nMJD - min(MJD) [d]" * " "^2, ylabel="",
    th_arrowlength=pi/3,
    panelbounds=true, ygrid=true,
    fontsizes=(thetalabel=9, ylabel=9, xlabel=9, thetaticklabel=9, xticklabel=9, yticklabel=7),
    thetaarrowkwargs=(color=:black, alpha=0.3),
    thetaticklabelkwargs=(halign=:center,),
    thetalabelkwargs=(halign=:center,),
    xtickkwargs=(linecolor=:black, linealpha=0.3,),
    xticklabelkwargs=(rotation=0, halign=:right, valign=:bottom),
    xlabelkwargs=(rotation=-90, halign=:right, valign=:top),
    ygridkwargs=(linecolor=:black, linealpha=0.3, linestyle=:solid,),
    yticklabelkwargs=(rotation=0,),
    # ylabelkwargs=(rotation=0, valign=:center, halign=:left),
    panelboundskwargs=(linecolor=:black, linealpha=0.5, linestyle=:solid,),
)
    
colors = palette(:rainbow, nthetas)

p = LVisP.plot(
    LVPC,
    theta_raw, x_raw, y_raw;
    yticks=yticks,
    thetaticklabels=["$thtl nm\nFluxcal []" for thtl in unique_thetas],
    plot_kwargs=[
    Dict(
        :mc=>colors[i], :label=>"",
        :seriestype=>:scatter,
    ) for i in eachindex(unique_thetas)]
)
plot!(p;
    size=(1200,1200),
    leftmargin=0Plots.mm, rightmargin=15Plots.mm,
    topmargin=15Plots.mm, bottommargin=0Plots.mm,
)

LVisP.plot!(
    LVPC,
    theta_pro, x_pro, y_pro;
    yticks=yticks,
    thetaticklabels=nothing,
    data_only=true,
    plot_kwargs=[Dict(
        :lc=>colors[i], :label=>"$(unique_thetas[i]) nm",
        :seriestype=>:path,
        :title=>"$otype ($survey)",
    ) for i in eachindex(unique_thetas)]
)


#traditional
p_trad1 = scatter(x_raw, y_raw;
    label="", mc=collect(colors)',
    xlabel="MJD-min(MJD) [d]",
    ylabel="Fluxcal []",
)
plot!(p_trad1, x_pro, y_pro; label=string.(theta_pro') .* " nm", lc=collect(colors)')

p_trad2 = plot(
    [plot(
        [x_raw[i], x_pro[i]], [y_raw[i], y_pro[i]];
        mc=colors[i], lc=colors[i],
        seriestype=[:scatter :path],
        label=["" "$(unique_thetas[i]) nm"],
        xlabel="MJD-min(MJD) [d]",
        ylabel="Fluxcal []",
        ylims=extrema(yticks),
    ) for i in eachindex(unique_thetas)]...
)

#combine LVisP and trad
p_comp = plot(p, p_trad1, p_trad2;
    size=(1200,900),
    layout=@layout[a [b;c]],
    rightmargin=5Plots.mm, leftmargin=-10Plots.mm,
    bottommargin=5Plots.mm,
)

display(p_comp)
savefig(p_comp, replace(fname, "./data/"=>"./gfx/", ".csv"=>".png", ".jl"=>".png"))

File: src_py_demo/LVisP_comparison.py

#%%imports
import glob
import importlib
import matplotlib.pyplot as plt
from matplotlib import colors as mcolors
import numpy as np
import os
import polars as pl
import re
import sys
from typing import Literal

sys.path.append("../")
from src_py import LVisPCanvas, utils as lvisu

importlib.reload(LVisPCanvas)
# plt.style.use("dark_background")

#%%definitions
def gaussian_pdf(x, mu, sigma):
    """
        - function defining a gaussian normal distribution
    """    
    y = 1/(sigma*np.sqrt(2*np.pi)) * np.exp(-(x - mu)**2 / (2*sigma**2))
    return y
def lc_sim(
    t:np.ndarray,
    t_peak:float, f_peak:float,
    stretch0:float, stretch1:float, stretch2:float,
    lbda:float=1.0,
    noiselevel:float=0.0,
    ) -> np.ndarray:
    """
        - function to define a very simplistic phenomenological LC simulation
    """
    f = (gaussian_pdf(t, t_peak - stretch0/2, stretch1) + gaussian_pdf(t, t_peak + stretch0/2, stretch2))
    f = f_peak * f / np.max(f) + noiselevel * np.random.randn(*t.shape)
    f *= lbda
    return f
def sin_sim(
    t:np.array,
    f_peak:float,
    p:float, offset:float=0.,
    noiselevel:float=0.0
    ) -> float:
    """
        - function to evaluate a sin with period `p` and `offset`
    """
    f = f_peak * np.sin(t * 2*np.pi/p + offset) + noiselevel * np.random.randn(*t.shape)
    return f
def simulate(
    nobjects:int=6,
    opt:Literal["lc","sin"]="lc"
    ):
    res = 500
    x = np.sort(np.random.choice(np.linspace(-50,100,res), size=(nobjects,res)), axis=1)
    theta_options = np.arange(0.2, 4, 0.2)
    theta = np.sort(np.random.choice(theta_options, size=nobjects, replace=False))
    
    if opt == "lc":
        t_peak = np.linspace(0,40,nobjects) * 0
        y           = np.array([*map(lambda i: lc_sim(x[i], t_peak=t_peak[i], f_peak=20, lbda=theta[i], stretch0=5, stretch1=15, stretch2=40, noiselevel=1.0), range(nobjects))])
        y_nonoise   = np.array([*map(lambda i: lc_sim(x[i], t_peak=t_peak[i], f_peak=20, lbda=theta[i], stretch0=5, stretch1=15, stretch2=40, noiselevel=0.0), range(nobjects))])
    else:
        theta *= 10
        y           = np.array([*map(lambda i: sin_sim(x[i], f_peak=1, p=theta[i], offset=0.0, noiselevel=0.1), range(nobjects))])
        y_nonoise   = np.array([*map(lambda i: sin_sim(x[i], f_peak=1, p=theta[i], offset=0.0, noiselevel=0.0), range(nobjects))])

    df_raw = pl.from_dict(dict(
            period=np.repeat(theta.reshape(-1,1), x.shape[1], axis=1).flatten(),
            time=x.flatten(),
            amplitude=y.flatten(),
            amplitude_e=np.nan,
            processing="raw",
        ))
    df_pro = pl.from_dict(dict(
            period=np.repeat(theta.reshape(-1,1), x.shape[1], axis=1).flatten(),
            time=x.flatten(),
            amplitude=y_nonoise.flatten(),
            amplitude_e=np.nan,
            processing="nonoise",
        ))
    
    df = pl.concat([df_raw, df_pro])
    

    return df

#%%data loading
#passbands
df_pb = pl.read_csv("../data/passband_specs.csv")
passbands = list(df_pb["name"])
pb_mappings = dict(zip(df_pb["wavelength"], df_pb.select(pl.exclude("wavelength")).to_numpy()))

#LCs
fnames = sorted(glob.glob("../data/*_elasticc.csv"))
fnames = np.append(fnames, ["../data/lc_simulated.py", "../data/sin_simulated.py"])
print(fnames)
fname = fnames[15]

#deal with on-the-fly data generation (pseudo filenames)
if fname == "../data/lc_simulated.py":
    df = simulate(9, opt="lc")
    for t in df["period"].unique(): pb_mappings[t] = [np.round(t, 3)]
    legend = False
    thetalab = "Maximum Scale"
    xlab = "Time [d]"
    ylab = "Amplitude []"
elif fname == "../data/sin_simulated.py":
    df = simulate(9, opt="sin")
    for t in df["period"].unique(): pb_mappings[t] = [np.round(t, 3)]
    legend = False
    thetalab = "Period [s]"
    xlab = "Time [s]"
    ylab = "Amplitude []"
else:
    df = pl.read_csv(fname, comment_prefix="#")
    legend = True
    thetalab = "Wavelength [nm]"
    xlab = "MJD-min(MJD) [d]" if "mjd" in df.columns else "Period [d]"
    ylab = "m [mag]" if "mag" in df.columns else "Fluxcal []"

# df = df.drop_nans()


parts = re.split(r"[/\_\.]", fname)
survey = parts[-2]
otype = parts[-3]

df_raw = df.filter(pl.col("processing")=="raw")
df_pro = df.filter(pl.col("processing")!="raw")
theta_raw = np.sort(np.unique(df_raw[:,0]))
df_raw_p = df_raw.partition_by(df_raw.columns[0], maintain_order=True)
x_raw = [df[:,1].to_numpy().astype(np.float64) for df in df_raw_p]
x_raw = [xi - np.nanmin(xi) for xi in x_raw]
y_raw = [df[:,2].to_numpy().astype(np.float64) for df in df_raw_p]
theta_pro = np.sort(np.unique(df_pro[:,0]))
df_pro_p = df_pro.partition_by(df_pro.columns[0], maintain_order=True)
x_pro = [df[:,1].to_numpy().astype(np.float64) for df in df_pro_p]
x_pro = [xi - np.nanmin(xi) for xi in x_pro]
y_pro = [df[:,2].to_numpy().astype(np.float64) for df in df_pro_p]

# print(theta_raw.shape, len(x_raw), len(y_raw))
# print(theta_pro.shape, len(x_pro), len(y_pro))

#%%get stats
unique_thetas = np.unique(theta_raw)
thetaticks = np.round(np.linspace(np.floor(np.min(theta_raw)), np.ceil(np.max(theta_raw)), 4),0).astype(int)
xticks = np.round(np.linspace(np.floor(np.min(np.concat(x_raw))), np.ceil(np.max(np.concat(x_raw))), 4), decimals=0).astype(int)
yticks = np.round(np.linspace(np.floor(np.min(np.concat(y_raw))), np.ceil(np.max(np.concat(y_raw))), 4), decimals=0).astype(int)
# yticks = np.sort(np.append(yticks, [-10, 80]))
panelsize = np.pi/10
vmin = 300 if ".py" not in fname else None
colors = lvisu.get_colors(theta_raw, cmap="nipy_spectral", vmin=vmin)

#%%plotting
#LVisP
fig = plt.figure(figsize=(12,9))
fig.suptitle(f"{otype} ({survey})")
ax = fig.add_subplot(121)
LVPC = LVisPCanvas.LVisPCanvas(ax,
    thetaticks, xticks, yticks,
    thetaguidelims=(-np.pi/2,np.pi/2), thetaplotlims=(-np.pi/2+panelsize/2,np.pi/2-panelsize/2),
    xlimdeadzone=0.3,
    thetalabel=thetalab, xlabel=xlab, ylabel=ylab,
    thetaarrowpos_th=None, ylabpos_th=None,
    thetatickkwargs=dict(c=plt.rcParams["text.color"]), thetaticklabelkwargs=None, thetalabelkwargs=None,
    xtickkwargs=None, xticklabelkwargs=dict(textcoords="offset fontsize", xytext=(-2,0)), xlabelkwargs=dict(rotation=-90,  textcoords="offset fontsize", xytext=(-3.5,0)),
    ylabelkwargs=dict(rotation=0)
)
LVPC.scatter(theta_raw, x_raw, y_raw,
    panel_kwargs=[dict(
        y_projection_method="y",
        panelsize=panelsize,
        show_panelbounds=True, show_yticks=True
    ) for _ in theta_raw],
    scatter_kwargs=[dict(
        c=mcolors.to_hex(colors[i]), label=pb_mappings[theta_raw[i]][0],
    ) for i in range(len(theta_raw))],
)
LVPC.plot(theta_pro, x_pro, y_pro,
    plot_kwargs=[dict(lw=3, c="w") for _ in theta_pro]
)
LVPC.plot(theta_pro, x_pro, y_pro,
    plot_kwargs=[dict(c=mcolors.to_hex(colors[i])) for i in range(len(theta_pro))],
)
# if legend: ax.legend()
ax.legend()


#traditional
axt1 = fig.add_subplot(222, xlabel=xlab, ylabel=ylab)

for i in range(len(theta_raw)):
    axt1.scatter(x_raw[i], y_raw[i], c=colors[i])
for i in range(len(theta_pro)):
    axt1.plot(x_pro[i], y_pro[i], c="w", lw=3)
    axt1.plot(x_pro[i], y_pro[i], c=colors[i], label=f"{pb_mappings[theta_raw[i]][0]} ({int(np.round(theta_raw[i], decimals=0))} nm)")
if legend: axt1.legend()

for i in range(len(theta_raw)):
# for i in range(6):
    ncols_theta = 3
    nrows_theta = int(np.ceil(len(theta_raw)/ncols_theta))
    nrows = int(2*nrows_theta)
    ncols = int(2*ncols_theta)
    pos = (1 + nrows_theta * ncols + ncols_theta) + i%ncols_theta + ncols*(i//ncols_theta)
    axt2 = fig.add_subplot(nrows, ncols, pos, xlabel=xlab, ylabel=ylab)
    # axt2 = fig.add_subplot(5, 1, i+1, xlabel="MJD-min(MJD) [d]", ylabel="Fuxcal []")
    axt2.scatter(x_raw[i], y_raw[i], c=colors[i])
    axt2.plot(x_pro[i], y_pro[i], c="w", lw=3)
    axt2.plot(x_pro[i], y_pro[i], c=colors[i], label=f"{pb_mappings[theta_raw[i]][0]} ({int(np.round(theta_raw[i], decimals=0))} nm)")
    if legend: axt2.legend()
    
fig.tight_layout()
fig.savefig(fname.replace("./data/","./gfx/").replace(".csv",".png").replace(".py",".png"), bbox_inches="tight")

plt.show()